src/ControlSystem/Trigger.hpp

```diff
@@ -83,125 +75,92 @@ class Trigger : public DenseTrigger {
-      // measurement_timescales are ready.
-      if (next_measurement(time, measurement_timescales) ==
-          std::numeric_limits<double>::infinity()) {
-        next_trigger_ = std::numeric_limits<double>::infinity();
-      } else {
-        next_trigger_ = time;
-      }
-    }
+      const control_system::FutureMeasurements& measurement_times) {
+    const auto next_measurement = measurement_times.next_measurement();
+    ASSERT(next_measurement.has_value(),
+           "Checking trigger without knowing next time.");
+    const bool triggered = time == *next_measurement;
 
     if (Parallel::get<Tags::Verbosity>(cache) >= ::Verbosity::Debug) {
       Parallel::printf(
           "%s, time = %.16f: Trigger for control systems (%s) is%s "
           "triggered.\n",
           get_output(array_index), time,
           pretty_type::list_of_names<ControlSystems>(),
-          (time == *next_trigger_ ? "" : " not"));
+          (triggered ? "" : " not"));
     }
 
-    return time == *next_trigger_;
+    return triggered;
   }
 
-  using next_check_time_return_tags = tmpl::list<>;
-  using next_check_time_argument_tags =
-      tmpl::list<::Tags::Time, control_system::Tags::MeasurementTimescales>;
+  using next_check_time_return_tags =
+      tmpl::list<control_system::Tags::FutureMeasurements<ControlSystems>>;
+  using next_check_time_argument_tags = tmpl::list<::Tags::Time>;
 
   template <typename Metavariables, typename ArrayIndex, typename Component>
   std::optional<double> next_check_time(
       Parallel::GlobalCache<Metavariables>& cache,
-      const ArrayIndex& array_index, const Component* component,
-      const double time,
-      const std::unordered_map<
-          std::string,
-          std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>>&
-          measurement_timescales) {
-    // At least one control system is active
-    const bool is_ready =
-        domain::functions_of_time_are_ready_algorithm_callback<
-            control_system::Tags::MeasurementTimescales>(
-            cache, array_index, component, time,
-            std::unordered_set{
-                control_system::combined_name<ControlSystems>()});
-    if (not is_ready) {
-      if (Parallel::get<Tags::Verbosity>(cache) >= ::Verbosity::Debug) {
-        Parallel::printf(
-            "%s, time = %.16f: Trigger - Cannot calculate next_check_time\n",
-            get_output(array_index), time);
-      }
-      return std::nullopt;
+      const ArrayIndex& array_index, const Component* /*component*/,
+      const gsl::not_null<control_system::FutureMeasurements*>
+          measurement_times,
+      const double time) {
+    if (measurement_times->next_measurement() == std::optional(time)) {
+      measurement_times->pop_front();
     }
 
-    const bool triggered = time == *next_trigger_;
-    if (triggered) {
-      *next_trigger_ = next_measurement(time, measurement_timescales);
+    if (not measurement_times->next_measurement().has_value()) {
```

Comment:
Will this work? Currently, every element has to wait for every FoT to be ready. Here, you're only checking the functions of time when these specific control systems need to be updated. But what about the other control systems? If we only check these specific FoTs on this element, then I believe your current implementation would work.